working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+
+  * using Dart as a container for shared business logic - which can be invoked on an arbitrary thread by surrounding native application. 
+
+  Native code does not _understand_ the concept of isolates. 
+
+The biggest challenge with introducing shared memory multithreading into Dart is avoiding breaking existing code, which is written under the assumption of single-threaded access. We would like to avoid just outright breaking existing code in subtle ways. Consider for a moment a library which is maintains a global cache internally:  
+
+```     dart
+library foo;
+
+class C {
+  final id;
+  C(this.id);
+}
+
+int _nextId;
+final _cache = <int, C>{};
+
+C makeObject() {
+  map[nextId] = C(nextId);
+  nextId++;
+  return map[nextId];
+}
+```
+
+This was a valid way to structure this code in single-threaded Dart, but the same code becomes thread _unsafe_ in the presence of shared memory multithreading.
+
+## Shareable Data
+
+We propose to extend Dart with a shared memory multithreading but provide a clear type-based delineation between two concurrency worlds. **Type of an instance should provide a clear indication whether its fields can be concurrently mutated by another thread.** We are going to call such types _shareable_ and assume that the language provides some way to opt-in into this for classes:
+
+```  dart
+/// Mark class sharable by implementing a [Shareable]
+/// interface.
+class S implements Shareable {
+  // ...
+} 
+```
+
+To maintain heap isolation we restrict the kind of data that a shareable class can contain: **Fields of sharable classes can only contain references to instances of sharable classes.** There is however no requirement of immutability imposed on shareable classes.
+
+References to shareable instances can be passed between isolates within an isolate group directly: for example sending a shareable instance through `SendPort` will not copy it.
+
+```dart  
+class S implements Shareable {
+  int v = 0;
+}
+
+final s = S();
+
+await Isolate.run(() {
+  // Even though we are running in a different isolate we have
+  // access to the original `s` and mutations performed in this
+  // isolate will be visible to other isolates.
+  // 
+  // (Subject to memory model constraints).
+  s.v = 10; 
+});
+
+expect(equals(10), s.v);
+```
+
+To make state sharing between isolates simpler we also allow to declare static fields which are shared between all isolates in the isolate group. `@shared` global fields are required to have static type that is a subtype of `Shareable?`. 
+
+```dart 
+// All isolates within an isolate group share this variable.
+@shared int v = 0;
+```
+
+### Functions
+
+Given the parallel between functions and other objects it makes sense to declare that functions which close only over shareable state are themselves shareable.
+
+**Static type of a function literal which captures only `Shareable` variables is  `ShareableFunction<F extends Function>` which is subtype of both `F` and `Shareable`.**
```

Comment:
> So, what if (strawman name) `ShareableAnd<T>` was the intersection type `T & Shareable` for any type, not just function types?

I like this idea. It could allow users to share classes from libraries that haven't opted in to this feature yet. They don't have to wait for the library developer to add `implements Shareable` to all their shareable classes.

Reply:
> It could allow users to share classes from libraries that haven't opted in to this feature yet. They don't have to wait for the library developer to add implements Shareable to all their shareable classes.

This might not be a good idea because methods on those classes might not be multithreading aware. 

I am inclined to think that introducing intersection types (especially with such a bizarre syntax) is too complicated. I am rewritting this section to remove `SharedFunction<T>` and make enforcement of this property dynamic or maybe not include shared functions into the initial version at all.